app/src/main/java/org/onionshare/android/tor/TorManager.kt

```diff
@@ -27,113 +31,149 @@ import java.io.FileInputStream
+        // bootstrapping gets 70% of our progress
+        if (keyword == EVENT_STATUS_CLIENT) {
+            val matchResult = BOOTSTRAP_REGEX.matchEntire(data)
+            val percent = matchResult?.groupValues?.get(1)?.toIntOrNull()
+            if (percent != null) {
+                val progress = (percent * 0.7).roundToInt()
+                val newState = (state.value as? TorState.Starting)?.copy(progress = progress)
+                    ?: TorState.Starting(progress)
+                _state.value = newState
+            }
+            return@RawEventListener
+        }
+        val onion = (state.value as? TorState.Starting)?.onion
+        // descriptor upload counts as 90%
+        if (onion != null && keyword == EVENT_HS_DESC && data.startsWith("UPLOAD $onion")) {
+            _state.value = TorState.Starting(90, onion)
+        }
+        // We consider already the first upload of the onion descriptor as started (100%).
+        // In practice, more uploads are needed for the onion service to be reachable.
+        if (onion != null && keyword == EVENT_HS_DESC && data.startsWith("UPLOADED $onion")) {
+            _state.value = TorState.Started(onion)
+        }
+    }
 
     /**
      * Starts [TorService] and creates a new onion service.
      * Suspends until the address of the onion service is available.
      */
-    suspend fun start(port: Int): String = suspendCoroutine { continuation ->
+    @Suppress("BlockingMethodInNonBlockingContext")
+    suspend fun start() = withContext(Dispatchers.IO) {
+        if (state.value !is TorState.Stopped) stop()
+
         LOG.info("Starting...")
-        broadcastReceiver = object : BroadcastReceiver() {
-            override fun onReceive(context: Context, i: Intent) {
-                when (i.getStringExtra(EXTRA_STATUS)) {
-                    TorService.STATUS_STARTING -> LOG.debug("TorService: Starting...")
-                    TorService.STATUS_ON -> {
-                        LOG.debug("TorService: Started")
-                        onTorStarted(context, continuation, port)
-                    }
-                    // FIXME When we stop unplanned, we need to inform the ShareManager
-                    //  that we stopped, so it can clear its state up, stopping webserver, etc.
-                    TorService.STATUS_STOPPING -> LOG.debug("TorService: Stopping...")
-                    TorService.STATUS_OFF -> LOG.debug("TorService: Stopped")
-                }
-            }
-        }
+        _state.value = TorState.Starting(0)
+        startLatch = CountDownLatch(1)
         app.registerReceiver(broadcastReceiver, IntentFilter(ACTION_STATUS))
+        broadcastReceiverRegistered = true
 
         Intent(app, ShareService::class.java).also { intent ->
             startForegroundService(app, intent)
         }
-        // this method suspends here until it the continuation resumes it
+        startLatch?.await() ?: error("startLatch was null")
```

Comment:
Before, I was using a `suspendCoroutine`, which isn't blocking, but suspending (subtle difference). This isn't working with a `final`/`val` broadcast receiver though as I need the continuation there to resume the coroutine. I guess blocking here is fine, rigiht? But in general, I am not yet completely happy yet with how things work here and in StateManager. Still very much MVP.

Reply:
Do you mean a suspendCoroutine suspends when it makes *any* call that would normally block? I thought blocking APIs had to be converted to suspend instead.

> I guess blocking here is fine, rigiht?

I have no idea. :-) Perhaps since we're on the IO dispatcher we're expected to make blocking IO calls, so it's fine to make other blocking calls too? But then why would anyone bother using coroutines on the IO dispatcher, if it's fine to just create a bunch of long-running tasks that block?

Reply:
Seems from [this article](https://elizarov.medium.com/blocking-threads-suspending-coroutines-d33e11bf4761) that yes, it's fine to block on the IO dispatcher.